CRM/Civicase/Form/Report/BaseExtendedReport.php

```diff
@@ -520,6 +524,222 @@ private function replaceNullRowValues(&$row, $key) {
     }
   }
 
+  /**
+   * Remove unused options for column.
+   *
+   * @param array $options
+   *   List of all column options.
+   * @param string $dbAlias
+   *   Db alias.
+   * @param string $fieldName
+   *   Field name.
+   */
+  private function removeUnusedOptions(array &$options, $dbAlias, $fieldName) {
+    if (!empty($this->_having)) {
+      $having = "{$this->_having} AND COUNT(*) > 0";
+    }
+    else {
+      $having = "HAVING COUNT(*) > 0";
+    }
+    $query = "SELECT {$dbAlias} {$this->_from} {$this->_where} GROUP BY {$dbAlias} {$having}";
+    $dao = CRM_Core_DAO::executeQuery($query);
+    $result = $dao->fetchAll();
+    $validOptions = [];
+    foreach ($result as $option) {
+      if (!empty($option[$fieldName])) {
+        if (strpos($option[$fieldName], CRM_Core_DAO::VALUE_SEPARATOR) !== FALSE) {
+          $multiOptions = explode(
+            CRM_Core_DAO::VALUE_SEPARATOR,
+            trim($option[$fieldName], CRM_Core_DAO::VALUE_SEPARATOR)
+          );
+          foreach ($multiOptions as $opt) {
+            $validOptions[$opt] = TRUE;
+          }
+        }
+        else {
+          $validOptions[$option[$fieldName]] = TRUE;
+        }
+      }
+    }
+
+    $options = array_intersect_key($options, $validOptions);
+  }
+
+  /**
+   * Build the report query.
+   *
+   * @param bool $applyLimit
+   *   Limit should be applied or not.
+   *
+   * @return string
+   *   Report query.
+   */
+  public function buildQuery($applyLimit = TRUE) {
+    if (empty($this->_params)) {
+      $this->_params = $this->controller->exportValues($this->_name);
+    }
+    $this->buildGroupTempTable();
+    $this->storeJoinFiltersArray();
+    $this->storeWhereHavingClauseArray();
+    $this->storeGroupByArray();
+    $this->storeOrderByArray();
+    $this->select();
+    $this->from();
+    $this->where();
+    $this->extendedCustomDataFrom();
+    $this->aggregateSelect();
+
+    if ($this->isInProcessOfPreconstraining()) {
+      $this->generateTempTable();
+      $this->_preConstrained = TRUE;
+      $this->select();
+      $this->from();
+      $this->extendedCustomDataFrom();
+      $this->constrainedWhere();
+      $this->aggregateSelect();
```

Comment:
Did we really need to override this function? I noticed the only change was just to re-order the function executions.

Reply:
Yes that reordering was necessary otherwise the tables for custom fields are not joined so I cannot reduce the options to used ones when user has selected the custom field as a column